ddtrace/settings/config.py

```diff
@@ -141,6 +159,229 @@ def get_error_ranges(error_range_str):
     return error_ranges  # type: ignore[return-value]
 
 
+_ConfigResult = TypedDict("_ConfigResult", {"value": Any, "raw": Any, "source": Any})
+
+
+class _ConfigSource(object):
+    """A configuration source that can be queried for a value."""
+
+    def get(self):
+        # type: () -> _ConfigResult
+        raise NotImplementedError
+
+
+class _ConfigSourceEnv(_ConfigSource):
+    """A configuration source that reads from an environment variable."""
+
+    def __init__(
+        self,
+        name,  # type: str
+        factory=lambda x: x,  # type: Callable[[str], Any]
+        examples=None,  # type: Optional[List[str]]
+    ):
+        # type: (...) -> None
+        self.key = name
+        self.factory = factory
+        self.examples = examples or []
+
+    def get(self):
+        # type: () -> _ConfigResult
+        raw = os.environ.get(self.key)
+        value = None
+        if raw is not None:
+            value = self.factory(raw)
+        return {"value": value, "raw": raw, "source": "env_var"}
+
+
+class _ConfigSourceEnvNone(_ConfigSourceEnv):
+    def __init__(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        pass
+
+    def get(self):
+        # type: () -> _ConfigResult
+        return {"value": None, "raw": None, "source": "env_var"}
+
+
+class _ConfigSourceEnvMulti(_ConfigSourceEnv):
+    def __init__(self, *envs, **kwargs):
+        # type: (*_ConfigSourceEnv, **Any) -> None
+        resolver = typing.cast("Optional[Callable[[_ConfigSourceEnv], _ConfigResult]]", kwargs.get("resolver"))
+        self.envs = envs
+        self.resolver = resolver or self._default_resolver
+
+    @staticmethod
+    def _default_resolver(*envs):
+        # type: (_ConfigSourceEnv) -> _ConfigResult
+        """
+        Default is to take the first env var that has a value.
+        """
+        for e in envs:
+            v = e.get()
+            if v["value"] is not None:
+                return v
+        return {
+            "value": None,
+            "raw": None,
+            "source": "env_var",
+        }
+
+    def get(self):
+        # type: () -> _ConfigResult
+        return self.resolver(*self.envs)
+
+
+class _ConfigSourceProgrammatic(_ConfigSource):
+    """A configuration source that is set programmatically by the user."""
+
+    def __init__(
+        self,
+        factory=None,  # type: Optional[Callable[[Any], Any]]
+    ):
+        # type: (...) -> None
+        self._value = None
+        self._raw = None
+        self._factory = factory
```

Comment:
What is the purpose of `self._factory` here? Is it to provide some processing of the raw input into a "correct" value? What happens when `self._factory` is None?

Reply:
Not sure if I'm misinterpreting, but is a `programmatic` config value stored directly in the `_ConfigSourceProgrammatic._value` object?